### PR TITLE
server: fix potential timezone related bugs caused by `time.Local`

### PR DESCRIPTION
**Pull request by Rustin-Liu**
### What problem does this PR solve? 
#13626

### Check List 

Tests 

None

Code changes

None

Side effects

None

Related changes

None

Release note

None

### CODE REVIEW

**Comment by Rustin-Liu** [comment]
@sre-bot /run-all-tests

**Comment by Rustin-Liu** [comment]
Pass all tests, PTAL~

**Comment by shenli** [comment]
@sre-bot /run-all-tests

**Comment by Rustin-Liu** [comment]
@sre-bot  /merge

**Comment by shenli** [comment]
@Rustin-Liu we are working on the CI problem.

**Comment by Rustin-Liu** [comment]
@sre-bot  /merge

**Comment by Rustin-Liu** [comment]
@sre-bot  /run-all-tests

**Comment by Rustin-Liu** [comment]
@sre-bot /run-all-tests

**Comment by Rustin-Liu** [comment]
@sre-bot /run-all-tests

**Comment by Rustin-Liu** [comment]
Hi @bb7133, it seem the CI broken because my code change.
Did you work no it?  Any update?

**Comment by bb7133** [comment]
@Rustin-Liu Very sorry for making it late, we'll try to fix the CI soon.

**Comment by Rustin-Liu** [comment]
> @Rustin-Liu Very sorry for making it late, we'll try to fix the CI soon.

Thank for your reply.  It's ok.

**Comment by reafans** [comment]
Hi, @Rustin-Liu, i checked the CI, and the reason why this pr can't pass the gorm-test is that the gorm would insert the default value 0001-01-01 00:00:00 to the TiDB, which would cause a timestamp {0 0 0 0 0 0 0} in `t.gotime()`, and the timestamp would turn into another timestamp in this method which cause an error. 
Before this pr, the `if t.Type() == mysql.TypeTimestamp && loc != nil` and `buffer, err = dumpBinaryDateTime(buffer, row.GetTime(i), nil)` would let the code pass, So if change `buffer, err = dumpBinaryDateTime(buffer, row.GetTime(i), loc)`  to `buffer, err = dumpBinaryDateTime(buffer, row.GetTime(i), nil)`. It will pass all the tests.

File: server/util.go [hunk 1]
```diff
@@ -263,7 +262,7 @@ func dumpBinaryRow(buffer []byte, columns []*ColumnInfo, row chunk.Row) ([]byte,
 			buffer = dumpLengthEncodedString(buffer, row.GetBytes(i))
 		case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
 			var err error
-			buffer, err = dumpBinaryDateTime(buffer, row.GetTime(i), nil)
+			buffer, err = dumpBinaryDateTime(buffer, row.GetTime(i), loc)
```
Comment (reafans) [inline server/util.go]:
```suggestion
			buffer, err = dumpBinaryDateTime(buffer, row.GetTime(i), nil)
```

File: server/util.go [hunk 2]
```diff
@@ -206,8 +206,7 @@ func dumpBinaryTime(dur time.Duration) (data []byte) {
 
 func dumpBinaryDateTime(data []byte, t types.Time, loc *time.Location) ([]byte, error) {
 	if t.Type() == mysql.TypeTimestamp && loc != nil {
```
Comment (reafans) [inline server/util.go]: before this pr, due to `loc != nil`, the `t1, err := t.GoTime(time.Local)` is useless.

**Comment by Rustin-Liu** [comment]
Hi @reafans , I think if pass the nil for this func then the code change will meaningless.
So I think we shouldn't do that.  Maybe we should reconsider the func logic to fix it.

**Comment by reafans** [comment]
Great! You can add some logic to filter the timestamp，and add more test for timestamp like "0000-00-00 00:00:00.0000000"，Thank you for your contribution!

**Comment by Rustin-Liu** [comment]
Hi @reafans , I think the actual reason of this test fail is 0001-01-01 00:00:00 not a valid time.
https://github.com/pingcap/tidb/blob/c7b43f3c75499efb07a8cba8949b20cdd7bed0a8/types/time.go#L1646

Maybe we should insert a valid time?

**Comment by Rustin-Liu** [comment]
ping~

**Comment by Rustin-Liu** [comment]
Hi @bb7133 @reafans, Any update?

**Comment by reafans** [comment]
Hi @Rustin-Liu ,the 0001-01-01 00:00:00 time is a default time when use gorm insert a time.Time field without initial, it can be convert into timestamp correctly with a session timezone in TiDB, so the code in line [#208](https://github.com/pingcap/tidb/pull/14572/files#diff-44da3d359afd2823a31e4fb048bdfdf4R208) - [#212](https://github.com/pingcap/tidb/pull/14572/files#diff-44da3d359afd2823a31e4fb048bdfdf4R212) is make no sense, it shouldn't be set a timezone when convert the resp data to binary in dumpBinaryRow(). This is why i advice you to change the loc to nil, or it would make the meaningless code be used and cause error.

**Comment by Rustin-Liu** [comment]
@sre-bot  /run-all-tests

**Comment by Rustin-Liu** [comment]
@sre-bot /run-all-tests

**Comment by Rustin-Liu** [comment]
@sre-bot /run-all-tests

**Comment by Rustin-Liu** [comment]
@sre-bot /run-all-tests

**Comment by Rustin-Liu** [comment]
Hi @reafans , Thanks for your reply.  I think if the loc always nil, we should remove the useless code.
PTAL~

**Comment by reafans** [comment]
Rest LGTM, something wrong with the integration-common-test, wait for the CI fixed.

**Comment by Rustin-Liu** [comment]
@sre-bot  /run-all-tests

**Comment by reafans** [comment]
lgtm

**Comment by Rustin-Liu** [comment]
@bb7133  cherry-pick failed.

**Pull request merged**